text/0000-item-grouping.md

```diff
@@ -0,0 +1,152 @@
+- Feature Name: `item_grouping`
+- Start Date: 2015-02-13
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Allow an anonymous `impl { ... }` for grouping items without introducing a module.
+
+# Motivation
+
+Many libraries today contain code like
+
+```rust
+#[cfg(target_os = "windows")]
+mod bar;
+
+#[cfg(target_os = "windows")]
+pub use bar::Baz;
+
+#[cfg(target_os = "windows")]
+fn quux() { ... }
+```
+
+With the new syntax, we can eliminate the duplication of `cfg` directives:
+
+```rust
+#[cfg(target_os = "windows")]
+impl {
+    mod bar;
+
+    pub use bar::Baz;
+
+    fn quux() { ... }
+}
+```
+
+This makes it easier to see at a glance which items are enabled when.  It's not
+a module, so you don't have to deal with another abstraction boundary or add a
+bunch more `use` items.
+
+Note that [`#[cfg]` on inherent `impl`s](http://is.gd/Jp2qVB) already works.
+
+Syntax extensions can similarly use this to process a set of items as a unit.
+For example I could rework [dynamodule][] to accept
+
+```rust
+#[class] impl {
+    struct Bicycle(&'static str);
+    type is_a = Vehicle;
+
+    fn new(color: &'static str) {
+        Bicycle(color)
+    }
+
+    fn go_somewhere(&self) -> String {
+        format!("{} bicycle has a basket and a bell that rings", self.color)
+    }
+}
+```
+
+Instead of digesting a whole function as token trees, we let libsyntax parse
+the `impl { ... }` normally, then fix up the AST at a higher level.
+(Naturally, this is only doable with procedural macros at the moment.)
+
+[dynamodule]: https://github.com/kmcallister/dynamodule#overview
+
+# Detailed design
+
+`impl { ... }` just provides a unit of grouping for `cfg`, syntax extensions,
+and similar purposes.  The items inside `impl { ... }` behave in every respect
+as though they were defined in the enclosing scope.
+
+(Of course, syntax extensions can perform arbitrary modifications to the AST.
+The `type is_a` above would *not* be defined in the enclosing scope, because
+the syntax extension interprets it specially.)
+
+The decision to support attributes besides `#[cfg]` can be made in the future
+on a case-by-cases basis.  With this PR alone,
+
+```rust
+#[derive(Clone)]
+impl {
+    struct Foo;
+    struct Bar;
+}
+```
```

Comment:
I've wanted this idea for a while.

---

If the `impl` follows the attribute, it seems less separated. It also seems quite arbitrary to allow some to be compact and others to not be.
```rust
#[derive(Clone)] impl {
    struct Foo;
    struct Bar;
}
```

Setting up phantom types for a bunch of units would be cleaner with some type of grouping. Verbose:
```rust
#[derive(Debug, Copy)]
struct Inch;
#[derive(Debug, Copy)]
struct Ft;
#[derive(Debug, Copy)]
struct Mile;

#[derive(Debug, Copy)]
struct Um; // Micrometer
#[derive(Debug, Copy)]
struct Mm;
#[derive(Debug, Copy)]
struct Cm;
#[derive(Debug, Copy)]
struct M;
#[derive(Debug, Copy)]
struct Km;

#[derive(Debug, Copy)]
struct Length<Unit, T>(T);
```

Compact:
```rust
#[derive(Debug, Copy)] impl {
    struct Inch;
    struct Ft;
    struct Mile;

    struct Um; // Micrometer
    struct Mm;
    struct Cm;
    struct M;
    struct Km;

    struct Length<Unit, T>(T);
}
```

There are obviously more unit types and this could be longer but I fail to see how restricting it from these is a good idea.